### PR TITLE
Check running job before create backup

### PR DESCRIPTION
**Pull request by pengweisong**
## What type of PR is this?
- [x] bug
- [ ] feature
- [ ] enhancement

#### Issue(s) number: 
https://github.com/vesoft-inc/nebula/issues/5012
https://github.com/vesoft-inc/nebula/issues/4975
## How do you solve it?
1. add more job type to check
2. add some test

## Checklist:
Tests:
- [ ] Unit test(positive and negative cases)
- [ ] Function test
- [ ] Performance test
- [ ] N/A

Affects:
- [ ] Documentation affected (Please add the label if documentation needs to be modified.)
- [ ] Incompatibility (If it breaks the compatibility, please describe it and add the label.）
- [ ] If it's needed to cherry-pick (If cherry-pick to some branches is required, please label the destination version(s).)
- [ ] Performance impacted: Consumes more CPU/Memory

## Release notes:

Please confirm whether to be reflected in release notes and how to describe:
> ex. Fixed the bug .....

### CODE REVIEW

File: src/meta/processors/admin/CreateBackupProcessor.cpp [hunk 1]
```diff
@@ -84,7 +84,11 @@ void CreateBackupProcessor::process(const cpp2::CreateBackupReq& req) {
...
   std::unordered_set<cpp2::JobType> jobTypes{cpp2::JobType::REBUILD_TAG_INDEX,
-                                             cpp2::JobType::REBUILD_EDGE_INDEX};
+                                             cpp2::JobType::REBUILD_EDGE_INDEX,
+                                             cpp2::JobType::COMPACT,
+                                             cpp2::JobType::INGEST,
+                                             cpp2::JobType::DATA_BALANCE,
+                                             cpp2::JobType::LEADER_BALANCE};
   auto result = jobMgr->checkTypeJobRunning(jobTypes);
   if (!nebula::ok(result)) {
     LOG(INFO) << "Get Index status failed, not allowed to create backup.";
```
Comment (panda-sheep) [inline src/meta/processors/admin/CreateBackupProcessor.cpp:94]: Maybe this log need to be changed？

**Review by panda-sheep — COMMENTED** [review]
Good job.
I have a little problem.
In the second step of `CreateBackupProcessor::process`, if a `blocking_` lock is added to the part of the storage engine, writing is not allowed, but it will not block the execution of the job? Such as compact job

File: src/interface/common.thrift [hunk 2]
```diff
@@ -403,7 +403,7 @@ enum ErrorCode {
     E_INVALID_JOB                     = -2065,  // Invalid task
 
     // Backup Failure
-    E_BACKUP_BUILDING_INDEX           = -2066,  // Backup terminated (index being created)
+    E_BACKUP_RUNNING_JOBS             = -2066,  // Backup terminated (index being created)
```
Comment (yixinglu) [inline src/interface/common.thrift:406]: plz update the error code comment and revisit the compatible issue caused by this thrift change

**Review by yixinglu — COMMENTED** [review]
Good enhancement!

**Comment by pengweisong** [comment]
> This PR is to prevent BR running when there are other jobs, which is definitely a must. But in the issue, it is the other way around: stop the Balance job running when there are backup jobs. Will it be solved somewhere else?

No, but since BR  use hard link instead of copy data, it will be very fast. But indeed, we may need sth to handle the corner case.

File: src/meta/processors/admin/CreateBackupProcessor.cpp [hunk 3]
```diff
@@ -93,8 +97,9 @@ void CreateBackupProcessor::process(const cpp2::CreateBackupReq& req) {
     return;
   }
   if (nebula::value(result)) {
-    LOG(INFO) << "Index is rebuilding, not allowed to create backup.";
-    handleErrorCode(nebula::cpp2::ErrorCode::E_BACKUP_BUILDING_INDEX);
+    LOG(INFO) << "Index rebuilding/Compact/Ingest/Balace job is running or queue, not allowed to "
```
[the remark this answers is not in the document]
Comment (pengweisong) [reply src/meta/processors/admin/CreateBackupProcessor.cpp:100]: It's hard to tell which type it is here, I'll try to log the specified type in the deeper code.

File: src/meta/processors/admin/CreateBackupProcessor.cpp (hunk 1 above)
Comment (pengweisong) [reply src/meta/processors/admin/CreateBackupProcessor.cpp:94]: indeed, thanks

**Comment by pengweisong** [comment]
> Good job. I have a question. In the second step of `CreateBackupProcessor::process`, if a `blocking_` lock is added to the part of the storage engine, writing is not allowed, but it will not block the execution of the job? Such as compact job

Ineed, then we should:
1. check when we start backup, there is no these jobs.
2. in the backup process we should prevent relevant jobs starting to run.

For the 1, we already check. The latter is the same problem raised by wenhao. We may could use an lock on space.

File: src/interface/common.thrift (hunk 2 above)
Comment (pengweisong) [reply src/interface/common.thrift:406]: good catch, thanks.

File: src/meta/processors/admin/CreateBackupProcessor.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (codesigner) [reply src/meta/processors/admin/CreateBackupProcessor.cpp:100]: plz try to print out which job is queued or running, block the backup, this LOG is not informative, do not let who read this log guess what is happening.

File: src/clients/meta/MetaClient.cpp [hunk 4]
```diff
@@ -933,8 +933,8 @@ Status MetaClient::handleResponse(const RESP& resp) {
       return Status::Error("Task report is out of date!");
     case nebula::cpp2::ErrorCode::E_BACKUP_FAILED:
       return Status::Error("Backup failure!");
-    case nebula::cpp2::ErrorCode::E_BACKUP_BUILDING_INDEX:
-      return Status::Error("Backup building indexes!");
+    case nebula::cpp2::ErrorCode::E_BACKUP_RUNNING_JOBS:
+      return Status::Error("Backup encounter running or queue jobs!");
```
Comment (codesigner) [inline src/clients/meta/MetaClient.cpp:937]: "queued jobs" looks better?

File: src/meta/processors/admin/CreateBackupProcessor.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (pengweisong) [reply src/meta/processors/admin/CreateBackupProcessor.cpp:100]: OK, fixed it

**Review by critical27 — DISMISSED** [review]

**Review by codesigner — APPROVED** [review]

**Review by critical27 — APPROVED** [review]

**Review by SuperYoko — APPROVED** [review]

**Review by yixinglu — DISMISSED** [review]

**Review by critical27 — APPROVED** [review]

**Review by SuperYoko — APPROVED** [review]

**Pull request merged**